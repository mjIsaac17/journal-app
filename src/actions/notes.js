import Swal from "sweetalert2";
import { db } from "../firebase/firebaseConfig";
import { fileUpload } from "../helpers/fileUpload";
import { loadNotes } from "../helpers/loadNotes";
import { types } from "../types/types";

export const startNewNote = () => {
  return async (dispatch, getState) => {
    //Video 260. getState returns all the state (auth, ui, notes)
    // const state = getState();
    const { uid } = getState().auth;
    const newNote = {
      title: "",
      body: "",
      date: new Date().getTime(),
    };
    try {
      //Order to store data in Firestore: collection/document/collection/document ...
      const doc = await db.collection(`${uid}/journal/notes`).add(newNote);
      // console.log(doc);
      dispatch(activeNote(doc.id, newNote));
      dispatch(addNewNoteInStore(doc.id, newNote));
    } catch (error) {
      console.log(error);
    }
  };
};

export const activeNote = (id, note) => ({
  type: types.notesActive,
  payload: {
    id,
    ...note,
  },
});

export const addNewNoteInStore = (id, note) => ({
  type: types.notesAddNew,
  payload: { id, ...note },
});

export const startLoadingNotes = (uid) => {
  return async (dispatch) => {
    const notes = await loadNotes(uid);
    dispatch(setNotes(notes));
  };
};

export const setNotes = (notes) => ({
  type: types.notesLoad,
  payload: notes,
});

export const startSaveNote = (note) => {
  //Because this is a sync task we use thunk to have access to the dispatch and getState
  return async (dispatch, getState) => {
    const { uid } = getState().auth;

    const noteToFireStore = { ...note };
    delete noteToFireStore.id;
    await db.doc(`${uid}/journal/notes/${note.id}`).update(noteToFireStore);
    dispatch(refreshNote(note.id, noteToFireStore));
    Swal.fire("Saved", note.title, "success");
  };
};

export const refreshNote = (id, note) => ({
  type: types.notesUpdate,
  payload: { id, note: { id, ...note } },
});

//If the task is async, the function name will begin with 'start'
export const startUploading = (file) => {
  return async (dispatch, getState) => {
    const { active: activeNote } = getState().notes;
    Swal.fire({
      title: "Uploading",
      text: "Please wait...",
      allowOutsideClick: false,
      // onBeforeOpen: () => {
      //   Swal.showLoading();
      // },
    });

    const fileUrl = await fileUpload(file);
    activeNote.url = fileUrl;
    dispatch(startSaveNote(activeNote));
    Swal.close();
  };
};
export const startDeleting = (id) => {
  return async (dispatch, getState) => {
    const uid = getState().auth.uid;
    //Delete note in Firebase
    await db.doc(`${uid}/journal/notes/${id}`).delete();

    dispatch(deleteNoteInStore(id));
  };
};

export const deleteNoteInStore = (id) => ({
  type: types.notesDelete,
  payload: id,
});

export const noteLogout = () => ({
  type: types.notesLogoutCleaning,
});
