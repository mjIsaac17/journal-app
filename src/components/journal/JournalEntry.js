import React from "react";

export const JournalEntry = () => {
  return (
    <div className="journal__entry pointer">
      <div
        className="journal__entry-picture"
        style={{
          backgroundSize: "cover",
          backgroundImage:
            "url(https://www.innovationnewsnetwork.com/wp-content/uploads/2020/12/space-junk-696x392.jpg)",
        }}
      ></div>
      <div className="journal__entry-body">
        <p className="journal__entry-title">New day</p>
        <p className="journal__entry-content">
          Lorem ipsum dolor sit amet, consectetur adipisicing elit.
        </p>
      </div>
      <div className="journal__entry-date-box">
        <span>Monday</span>
        <h4>26</h4>
      </div>
    </div>
  );
};
