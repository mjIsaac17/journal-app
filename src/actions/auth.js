import { firebase, googleAuthProvider } from "../firebase/firebaseConfig";
import Swal from "sweetalert2";
import { types } from "../types/types";
import { finishLoading, startLoading } from "./ui";
import { noteLogout } from "./notes";

export const startLoginEmailPassword = (email, password) => {
  //this dispatch is offered by thunk
  return (dispatch) => {
    dispatch(startLoading());
    firebase
      .auth()
      .signInWithEmailAndPassword(email, password)
      .then(({ user }) => {
        dispatch(finishLoading());
        dispatch(login(user.uid, user.displayName));
      })
      .catch((err) => {
        dispatch(finishLoading());
        Swal.fire("Error", err.message, "error");
        // console.log(err);
      });
  };
};

export const startRegisterWithEmailPasswordName = (email, password, name) => {
  return (dispatch) => {
    firebase
      .auth()
      .createUserWithEmailAndPassword(email, password)
      .then(async ({ user }) => {
        //We use async-await to avoid adding another then
        //Update the user name from null to "name".
        await user.updateProfile({ displayName: name });
        // console.log(user);
        dispatch(login(user.uid, user.displayName));
      })
      .catch((err) => {
        Swal.fire("Error", err.message, "error");
        // console.log(err);
      });
  };
};

export const startGoogleLogin = () => {
  return (dispatch) => {
    firebase
      .auth()
      .signInWithPopup(googleAuthProvider)
      .then(({ user }) => {
        dispatch(login(user.uid, user.displayName));
      });
  };
};

export const login = (uid, displayName) => ({
  //implicit return
  type: types.login,
  payload: {
    uid,
    displayName,
  },
});

export const startLogout = () => {
  return async (dispatch) => {
    await firebase.auth().signOut();
    dispatch(logout());
    dispatch(noteLogout());
  };
};

export const logout = () => ({
  type: types.logout,
});
