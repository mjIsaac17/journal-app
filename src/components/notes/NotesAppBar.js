import React from "react";
import { useDispatch, useSelector } from "react-redux";
import { startSaveNote, startUploading } from "../../actions/notes";

export const NotesAppBar = () => {
  const dispatch = useDispatch();
  const note = useSelector((state) => state.notes.active);
  const handleSave = () => {
    if (!note.url) {
      delete note.url;
    }
    dispatch(startSaveNote(note));
    //console.log(note);
  };

  const handlePictureClick = () => {
    document.querySelector("#iInputFile").click();
  };
  const handleFileChange = (e) => {
    const file = e.target.files[0];
    if (file) {
      dispatch(startUploading(file));
    }
  };
  return (
    <div className="notes__appbar">
      <span>September 17th 2021</span>
      <input
        id="iInputFile"
        type="file"
        style={{ display: "none" }}
        onChange={handleFileChange}
      />
      <div>
        <button className="btn" onClick={handlePictureClick}>
          Picture
        </button>
        <button className="btn" onClick={handleSave}>
          Save
        </button>
      </div>
    </div>
  );
};
