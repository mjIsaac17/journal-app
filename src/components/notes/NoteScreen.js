import React from "react";
import { useRef } from "react";
import { useEffect } from "react";
import { useDispatch, useSelector } from "react-redux";
import { activeNote } from "../../actions/notes";
import { useForm } from "../../hooks/useForm";
import { NotesAppBar } from "./NotesAppBar";

export const NoteScreen = () => {
  const dispatch = useDispatch();
  const { active: note } = useSelector((state) => state.notes);
  const [formValues, handleInputChange, reset] = useForm(note);
  const { body, title } = formValues;

  const activeId = useRef(note.id);

  useEffect(() => {
    if (note.id !== activeId.current) {
      reset(note);
      activeId.current = note.id;
    }
  }, [note, reset]);

  useEffect(() => {
    //Update the state when we change the text values of the note
    dispatch(activeNote(formValues.id, { ...formValues }));
  }, [formValues, dispatch]);
  return (
    <div className="notes__main-content">
      <NotesAppBar />
      <div className="notes__content">
        <input
          value={title}
          type="text"
          name="title" //we need to add the name otherwise the text will not change. This is needed in useForm
          placeholder="Some awesome title"
          className="notes__title-input"
          onChange={handleInputChange}
        />
        <textarea
          value={body}
          name="body"
          placeholder="What happened today?"
          className="notes__textarea"
          onChange={handleInputChange}
        ></textarea>
        <div className="notes__image">
          <img
            src="https://encrypted-tbn0.gstatic.com/images?q=tbn:ANd9GcR4d0Yi8SwQOlbmnL0BJzjGbI5N8LU1Omdz2Q&usqp=CAU"
            alt="The Moon"
          />
        </div>
      </div>
    </div>
  );
};
