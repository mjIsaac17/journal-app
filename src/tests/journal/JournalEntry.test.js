import { mount } from "enzyme";
import { Provider } from "react-redux";

import configureStore from "redux-mock-store";
import thunk from "redux-thunk";
import { JournalEntry } from "../../components/journal/JournalEntry";
import { activeNote } from "../../actions/notes";

// jest.mock("../../actions/notes", () => ({
//   activeNote: jest.fn(),
// }));

const middlewares = [thunk];
const mockStore = configureStore(middlewares);

const initState = {
  auth: {},
  ui: {
    loading: false,
    msgError: null,
  },
};

let store = mockStore({});
store.dispatch = jest.fn();
describe("Tests in <JournalEntry />", () => {
  const note = {
    id: "123",
    title: "This is a title",
    body: "React course",
    date: 123,
    url: "image.jpg",
  };
  const wrapper = mount(
    <Provider store={store}>
      <JournalEntry {...note} />
    </Provider>
  );
  test("should load the component properly", () => {
    expect(wrapper).toMatchSnapshot();
  });

  // 1. Option to test activeNote, it is needed to use the mock
  //   test("should trigger activeNote with the selected note", () => {
  //     wrapper.find(".journal__entry").prop("onClick")();
  //     expect(activeNote).toHaveBeenCalledWith(note.id, note);
  //   });

  // 2. Other alternative to the activeNote with the store
  test("should active the note", () => {
    wrapper.find(".journal__entry").prop("onClick")();
    //store.dispatch contains the arguments with which the function was called
    expect(store.dispatch).toHaveBeenCalledWith(activeNote("1", note));
  });
});
