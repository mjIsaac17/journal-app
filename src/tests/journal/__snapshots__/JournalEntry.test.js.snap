// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tests in <JournalEntry /> should load the component properly 1`] = `
<div
  className="journal__entry pointer animate__animated animate__fadeIn animate__faster"
  onClick={[Function]}
>
  <div
    className="journal__entry-picture"
    style={
      Object {
        "backgroundImage": "url(image.jpg)",
        "backgroundSize": "cover",
      }
    }
  />
  <div
    className="journal__entry-body"
  >
    <p
      className="journal__entry-title"
    >
      This is a title
    </p>
    <p
      className="journal__entry-content"
    >
      React course
    </p>
  </div>
  <div
    className="journal__entry-date-box"
  >
    <span>
      Wednesday
    </span>
    <h4>
      31st
    </h4>
  </div>
</div>
`;
